compass/ocean/tests/nonhydro/stratified_seiche/initial_state.py

```diff
@@ -0,0 +1,221 @@
+        layerThickness[:] = -1e34
+
+        # equally spaced layers
+        refLayerThickness[:] = maxDepth / nVertLevels
+        refBottomDepth[0] = refLayerThickness[0]
+        refZMid[0] = -0.5 * refLayerThickness[0]
+        for k in range(1, nVertLevels):
+            refBottomDepth[k] = refBottomDepth[k - 1] + refLayerThickness[k]
+            refZMid[k] = -refBottomDepth[k - 1] - 0.5 * refLayerThickness[k]
+
+        # SSH
+        ssh[:] = 0.0
+
+        # Compute maxLevelCell and layerThickness for z-level
+        # (variation only on top)
+        if (vertical_coordinate == 'z'):
+            vertCoordMovementWeights[:] = 0.0
+            vertCoordMovementWeights[0] = 1.0
+            for iCell in range(0, nCells):
+                maxLevelCell[iCell] = nVertLevels - 1
+                bottomDepth[iCell] = refBottomDepth[nVertLevels - 1] 
+                layerThickness[0, iCell, :] = refLayerThickness[:]
+                layerThickness[0, iCell, 0] += ssh[iCell]
+            restingThickness[:, :] = layerThickness[0, :, :]
+        # Compute maxLevelCell and layerThickness for uniform
+        elif (vertical_coordinate == 'uniform'):
+            vertCoordMovementWeights[:] = 1.0
+            vertCoordMovementWeights[0] = 1.0
+            for iCell in range(0, nCells):
+                maxLevelCell[iCell] = nVertLevels - 1
+                bottomDepth[iCell] = refBottomDepth[nVertLevels - 1]
+                layerThickness[0, iCell, :] = refLayerThickness[:] + \
+                    ssh[iCell]/nVertLevels
+            restingThickness[:, :] = refLayerThickness[:]
+
+        # Compute zMid (same, regardless of vertical coordinate)
+        for iCell in range(0, nCells):
+            k = maxLevelCell[iCell]
+            zMid[0, iCell, k] = -bottomDepth[iCell] + \
+                0.5 * layerThickness[0, iCell, k]
+            for k in range(maxLevelCell[iCell] - 1, -1, -1):
+                zMid[0, iCell, k] = zMid[0, iCell, k + 1] + 0.5 * \
+                    (layerThickness[0, iCell, k + 1] + \
+                    layerThickness[0, iCell, k])
+
+        # initialize tracers
+        rho0 = 1000.0  # kg/m^3
+        rhoz = -2.0e-4  # kg/m^3/m in z
+        S0 = 35.0
+        r = 1
+
+        # linear equation of state
+        # rho = rho0 - alpha*(T-Tref) + beta*(S-Sref)
+        # set S=Sref
+        # T = Tref - (rho - rhoRef)/alpha
+        for k in range(0, nVertLevels):
+            activeCells = k <= maxLevelCell
+            salinity[0, activeCells, k] = S0 
+            density[0, activeCells, k] = rho0 + (10.0/2)*(1.0 - \
+                np.tanh((2/1)*np.arctanh(0.99)*(zMid[0, :, k] + \
+                0.5*maxDepth - 0.1*np.cos((np.pi/10)*xCell[:]))))
+            # T = Tref - (rho - rhoRef)/alpha
+            temperature[0, activeCells, k] = config_eos_linear_Tref \
+                - (density[0, activeCells, k] - config_eos_linear_densityref) / \
+                config_eos_linear_alpha
```

Comment:
It would be better to avoid looping over k if we don't need to.  This would have at least advantages:
1. you would not need to pre-initialized the 3D variables above
2. looping is nearly always slow in python

The computation of the vertical coordinate that I suggested above will add a `ds.cellMask` variable that I demonstrate here (equivalent to your `activeCells`). But in this test case, you have hard-coded that there is no bathymetry so this might be needlessly complicated.  You might want to leave out the masking (all `.where(ds.cellMask)` calls).

```suggestion
        
        ds['salinity'] = (S0*xarray.ones_like(ds.zMid)).where(ds.cellMask)
        ds['density'] =rho0 + (10.0/2)*(1.0 - \
                np.tanh((2/1)*np.arctanh(0.99)*(ds.zMid + \
                0.5*maxDepth - 0.1*np.cos((np.pi/10)*ds.xCell))))
            # T = Tref - (rho - rhoRef)/alpha
            ds['temperature'] =  config_eos_linear_Tref \
                - (ds.density - config_eos_linear_densityref) / \
                config_eos_linear_alpha
```
The density calculations seems to be really tricky to format in a reasonable way so maybe compute some intermediate variable to make it more readable?

Reply:
Changed it as suggested, but I have not put intermediate variables. If you think they are necessary for the code to be readable I'll create them.